File: sdk/typescript/tests/process_logs.test.ts
import express from 'express';
import type { AddressInfo } from 'node:net';
import { afterEach, describe, expect, it } from 'vitest';

import { ProcessLogRing, registerAgentfieldLogsRoute } from '../src/agent/processLogs.js';

const ENV_KEYS = [
  'AGENTFIELD_LOGS_ENABLED',
  'AGENTFIELD_LOG_BUFFER_BYTES',
  'AGENTFIELD_LOG_MAX_LINE_BYTES',
  'AGENTFIELD_LOG_MAX_TAIL_LINES',
  'AGENTFIELD_AUTHORIZATION_INTERNAL_TOKEN'
] as const;

afterEach(() => {
  for (const key of ENV_KEYS) {
    delete process.env[key];
  }
});

function applyEnv(overrides: Partial<Record<(typeof ENV_KEYS)[number], string>>): void {
  for (const key of ENV_KEYS) {
    delete process.env[key];
  }
  Object.assign(process.env, overrides);
}

function parseNdjson(text: string): Array<Record<string, unknown>> {
  return text
    .trim()
    .split('\n')
    .filter(Boolean)
    .map((line) => JSON.parse(line) as Record<string, unknown>);
}

async function withLogsServer(
  ring: ProcessLogRing,
  callback: (baseUrl: string) => Promise<void>
): Promise<void> {
  const app = express();
  registerAgentfieldLogsRoute(app, ring);

  const server = await new Promise<ReturnType<typeof app.listen>>((resolve) => {
    const instance = app.listen(0, () => resolve(instance));
  });

  try {
    const address = server.address() as AddressInfo;
    await callback(`http://127.0.0.1:${address.port}`);
  } finally {
    server.closeIdleConnections?.();
    server.closeAllConnections?.();
    await new Promise<void>((resolve, reject) => {
      server.close((error?: Error) => (error ? reject(error) : resolve()));
    });
  }
}

describe('ProcessLogRing', () => {
  it('tails and snapshots logs while trimming to the configured buffer size', () => {
    applyEnv({ AGENTFIELD_LOG_BUFFER_BYTES: '1024' });
    const ring = new ProcessLogRing();
    const longLine = 'x'.repeat(400);

    ring.append('stdout', `${longLine}-first`, false);
    ring.append('stderr', `${longLine}-second`, true);
    ring.append('custom', `${longLine}-third`, false);

    expect(ring.tail(0)).toEqual([]);

    const entries = ring.tail(10);
    expect(entries).toHaveLength(2);
    expect(entries.map((entry) => entry.line)).toEqual([
      `${longLine}-second`,
      `${longLine}-third`
    ]);
    expect(entries[0]?.level).toBe('error');
    expect(entries[0]?.truncated).toBe(true);
    expect(entries[1]?.level).toBe('log');

    expect(ring.snapshotAfter(1, null).map((entry) => entry.line)).toEqual([
      `${longLine}-second`,
      `${longLine}-third`
    ]);
    expect(ring.snapshotAfter(1, 1).map((entry) => entry.line)).toEqual([`${longLine}-third`]);
  });
});

describe('registerAgentfieldLogsRoute', () => {
  it('returns 404 when process logs are disabled', async () => {
    applyEnv({ AGENTFIELD_LOGS_ENABLED: 'false' });
    const ring = new ProcessLogRing();

    await withLogsServer(ring, async (baseUrl) => {
      const response = await fetch(`${baseUrl}/agentfield/v1/logs`);

      expect(response.status).toBe(404);
      await expect(response.json()).resolves.toEqual({
        error: 'logs_disabled',
        message: 'Process logs API is disabled'
      });
    });
  });

  it('enforces the internal bearer token and tail_lines cap', async () => {
    applyEnv({
      AGENTFIELD_LOGS_ENABLED: 'true',
      AGENTFIELD_AUTHORIZATION_INTERNAL_TOKEN: 'secret-token',
      AGENTFIELD_LOG_MAX_TAIL_LINES: '1'
    });
    const ring = new ProcessLogRing();
    ring.append('stdout', 'first line', false);

    await withLogsServer(ring, async (baseUrl) => {
      const unauthorized = await fetch(`${baseUrl}/agentfield/v1/logs`);
      expect(unauthorized.status).toBe(401);
      await expect(unauthorized.json()).resolves.toEqual({
        error: 'unauthorized',
        message: 'Valid Authorization Bearer required'
      });

      const tooLarge = await fetch(`${baseUrl}/agentfield/v1/logs?tail_lines=2`, {
        headers: { Authorization: 'Bearer secret-token' }
      });
      expect(tooLarge.status).toBe(413);
      await expect(tooLarge.json()).resolves.toEqual({
        error: 'tail_too_large',
        message: 'tail_lines exceeds max 1'
      });
    });
  });

  it('returns ndjson tail output and supports since_seq filtering', async () => {
    applyEnv({
      AGENTFIELD_LOGS_ENABLED: 'true',
      AGENTFIELD_AUTHORIZATION_INTERNAL_TOKEN: 'secret-token'
    });
    const ring = new ProcessLogRing();
    ring.append('stdout', 'first line', false);
    ring.append('stderr', 'second line', true);

    await withLogsServer(ring, async (baseUrl) => {
      const response = await fetch(`${baseUrl}/agentfield/v1/logs`, {
        headers: { Authorization: 'Bearer secret-token' }
      });

      expect(response.status).toBe(200);
      expect(response.headers.get('content-type')).toContain('application/x-ndjson');
      expect(response.headers.get('cache-control')).toBe('no-store');

      const initial = parseNdjson(await response.text());
      expect(initial.map((entry) => entry.line)).toEqual(['first line', 'second line']);
      expect(initial[1]?.level).toBe('error');
      expect(initial[1]?.truncated).toBe(true);

      const sinceResponse = await fetch(`${baseUrl}/agentfield/v1/logs?since_seq=1&tail_lines=1`, {
        headers: { Authorization: 'Bearer secret-token' }
      });

      expect(sinceResponse.status).toBe(200);
      expect(parseNdjson(await sinceResponse.text()).map((entry) => entry.line)).toEqual([
        'second line'
      ]);
    });
  });
});

File: sdk/go/agent/process_logs_test.go
package agent

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func testAgentWithProcessLogRing(ring *processLogRing) *Agent {
	a := &Agent{procLogRing: ring}
	a.procLogOnce.Do(func() {})
	return a
}

func decodeProcessLogEntries(t *testing.T, body string) []processLogEntry {
	t.Helper()
	trimmed := strings.TrimSpace(body)
	if trimmed == "" {
		return nil
	}
	lines := strings.Split(trimmed, "\n")
	entries := make([]processLogEntry, 0, len(lines))
	for _, line := range lines {
		if strings.TrimSpace(line) == "" {
			continue
		}
		var entry processLogEntry
		if err := json.Unmarshal([]byte(line), &entry); err != nil {
			t.Fatalf("unmarshal process log entry: %v", err)
		}
		entries = append(entries, entry)
	}
	return entries
}

func TestProcessLogs_Ring(t *testing.T) {
	t.Run("config parsing and auth", func(t *testing.T) {
		t.Setenv("AGENTFIELD_LOG_BUFFER_BYTES", "invalid")
		if got := processLogsMaxBytes(); got != 4<<20 {
			t.Fatalf("processLogsMaxBytes invalid = %d", got)
		}

		t.Setenv("AGENTFIELD_LOG_BUFFER_BYTES", "2048")
		if got := processLogsMaxBytes(); got != 2048 {
			t.Fatalf("processLogsMaxBytes = %d", got)
		}

		t.Setenv("AGENTFIELD_LOG_MAX_LINE_BYTES", "100")
		if got := processLogsMaxLineBytes(); got != 16384 {
			t.Fatalf("processLogsMaxLineBytes invalid = %d", got)
		}

		t.Setenv("AGENTFIELD_LOG_MAX_LINE_BYTES", "512")
		if got := processLogsMaxLineBytes(); got != 512 {
			t.Fatalf("processLogsMaxLineBytes = %d", got)
		}

		t.Setenv("AGENTFIELD_LOG_MAX_TAIL_LINES", "0")
		if got := processLogsMaxTailLines(); got != 50000 {
			t.Fatalf("processLogsMaxTailLines invalid = %d", got)
		}

		t.Setenv("AGENTFIELD_LOG_MAX_TAIL_LINES", "12")
		if got := processLogsMaxTailLines(); got != 12 {
			t.Fatalf("processLogsMaxTailLines = %d", got)
		}

		t.Setenv("AGENTFIELD_LOGS_ENABLED", "off")
		if processLogsEnabled() {
			t.Fatal("expected process logs disabled")
		}

		t.Setenv("AGENTFIELD_LOGS_ENABLED", "true")
		if !processLogsEnabled() {
			t.Fatal("expected process logs enabled")
		}

		t.Setenv("AGENTFIELD_AUTHORIZATION_INTERNAL_TOKEN", "secret")
		if internalBearerOK("Bearer wrong") {
			t.Fatal("expected bearer mismatch to fail")
		}
		if !internalBearerOK("Bearer secret") {
			t.Fatal("expected bearer token to pass")
		}
	})

	t.Run("ring trimming and snapshots", func(t *testing.T) {
		if ring := newProcessLogRing(1); ring.maxBytes != 1024 {
			t.Fatalf("newProcessLogRing min bytes = %d", ring.maxBytes)
		}

		var nilRing *processLogRing
		nilRing.appendLine("stdout", "ignored", false)

		ring := newProcessLogRing(1024)
		longLine := strings.Repeat("x", 400)
		ring.appendLine("stdout", longLine+"-first", false)
		ring.appendLine("stderr", longLine+"-second", true)
		ring.appendLine("custom", longLine+"-third", false)

		if got := ring.tail(0); got != nil {
			t.Fatalf("tail(0) = %#v", got)
		}

		entries := ring.tail(10)
		if len(entries) != 2 {
			t.Fatalf("tail entries = %d", len(entries))
		}
		if entries[0].Level != "error" || !entries[0].Truncated {
			t.Fatalf("unexpected stderr entry: %#v", entries[0])
		}
		if entries[1].Level != "log" {
			t.Fatalf("unexpected custom level: %#v", entries[1])
		}

		since := ring.snapshotAfter(1, 0)
		if len(since) != 2 {
			t.Fatalf("snapshotAfter len = %d", len(since))
		}
		limited := ring.snapshotAfter(1, 1)
		if len(limited) != 1 || !strings.HasSuffix(limited[0].Line, "-third") {
			t.Fatalf("snapshotAfter limit = %#v", limited)
		}
	})
}

func TestProcessLogs_Handler(t *testing.T) {
	t.Run("method not allowed", func(t *testing.T) {
		t.Setenv("AGENTFIELD_LOGS_ENABLED", "true")
		req := httptest.NewRequest(http.MethodPost, "/agentfield/v1/logs", nil)
		resp := httptest.NewRecorder()

		testAgentWithProcessLogRing(newProcessLogRing(1024)).handleAgentfieldLogs(resp, req)

		if resp.Code != http.StatusMethodNotAllowed {
			t.Fatalf("status = %d", resp.Code)
		}
	})

	t.Run("disabled and unauthorized responses", func(t *testing.T) {
		t.Setenv("AGENTFIELD_LOGS_ENABLED", "false")
		req := httptest.NewRequest(http.MethodGet, "/agentfield/v1/logs", nil)
		resp := httptest.NewRecorder()

		testAgentWithProcessLogRing(newProcessLogRing(1024)).handleAgentfieldLogs(resp, req)

		if resp.Code != http.StatusNotFound || !strings.Contains(resp.Body.String(), "logs_disabled") {
			t.Fatalf("unexpected disabled response: %d %s", resp.Code, resp.Body.String())
		}

		t.Setenv("AGENTFIELD_LOGS_ENABLED", "true")
		t.Setenv("AGENTFIELD_AUTHORIZATION_INTERNAL_TOKEN", "secret")
		unauthorized := httptest.NewRecorder()
		testAgentWithProcessLogRing(newProcessLogRing(1024)).handleAgentfieldLogs(unauthorized, req)
		if unauthorized.Code != http.StatusUnauthorized || !strings.Contains(unauthorized.Body.String(), "unauthorized") {
			t.Fatalf("unexpected unauthorized response: %d %s", unauthorized.Code, unauthorized.Body.String())
		}
	})

	t.Run("tailing and follow mode", func(t *testing.T) {
		t.Setenv("AGENTFIELD_LOGS_ENABLED", "true")
		t.Setenv("AGENTFIELD_AUTHORIZATION_INTERNAL_TOKEN", "secret")
		t.Setenv("AGENTFIELD_LOG_MAX_TAIL_LINES", "1")

		ring := newProcessLogRing(1024)
		ring.appendLine("stdout", "first line", false)
		ring.appendLine("stderr", "second line", true)
		a := testAgentWithProcessLogRing(ring)

		tooLargeReq := httptest.NewRequest(http.MethodGet, "/agentfield/v1/logs?tail_lines=2", nil)
		tooLargeReq.Header.Set("Authorization", "Bearer secret")
		tooLargeResp := httptest.NewRecorder()
		a.handleAgentfieldLogs(tooLargeResp, tooLargeReq)
		if tooLargeResp.Code != http.StatusRequestEntityTooLarge || !strings.Contains(tooLargeResp.Body.String(), "tail_too_large") {
			t.Fatalf("unexpected tail too large response: %d %s", tooLargeResp.Code, tooLargeResp.Body.String())
		}

		defaultReq := httptest.NewRequest(http.MethodGet, "/agentfield/v1/logs", nil)
		defaultReq.Header.Set("Authorization", "Bearer secret")
		defaultResp := httptest.NewRecorder()
		a.handleAgentfieldLogs(defaultResp, defaultReq)
		if defaultResp.Code != http.StatusOK {
			t.Fatalf("status = %d", defaultResp.Code)
		}
		if got := defaultResp.Header().Get("Cache-Control"); got != "no-store" {
			t.Fatalf("cache-control = %q", got)
		}
		entries := decodeProcessLogEntries(t, defaultResp.Body.String())
		if len(entries) != 2 || entries[0].Line != "first line" || entries[1].Line != "second line" {
			t.Fatalf("default entries = %#v", entries)
		}

		ctx, cancel := context.WithCancel(context.Background())
		cancel()
		followReq := httptest.NewRequest(http.MethodGet, "/agentfield/v1/logs?since_seq=1&tail_lines=1&follow=true", nil).WithContext(ctx)
		followReq.Header.Set("Authorization", "Bearer secret")
		followResp := httptest.NewRecorder()
		a.handleAgentfieldLogs(followResp, followReq)
		followEntries := decodeProcessLogEntries(t, followResp.Body.String())
		if len(followEntries) != 1 || followEntries[0].Line != "second line" {
			t.Fatalf("follow entries = %#v", followEntries)
		}
	})
	

t.Run("evicts oldest entries when exceeding maxBytes", func(t *testing.T) {
	ring := newProcessLogRing(100) // VERY small buffer

	big := strings.Repeat("x", 500) // VERY large lines

	ring.appendLine("stdout", big+"1", false)
	ring.appendLine("stdout", big+"2", false)
	ring.appendLine("stdout", big+"3", false)

	entries := ring.tail(10)

	// After eviction, only latest entry should remain
	if len(entries) != 1 {
		t.Fatalf("expected 1 entry after eviction, got %d", len(entries))
	}

	if !strings.HasSuffix(entries[0].Line, "3") {
		t.Fatalf("expected latest entry to remain, got %#v", entries)
	}
})
    t.Run("snapshotAfter returns only newer entries", func(t *testing.T) {
	ring := newProcessLogRing(1024)

	ring.appendLine("stdout", "first", false)
	ring.appendLine("stdout", "second", false)
	ring.appendLine("stdout", "third", false)

	entries := ring.snapshotAfter(2, 0)

	if len(entries) != 1 || entries[0].Line != "third" {
		t.Fatalf("unexpected snapshotAfter result: %#v", entries)
	}
   })
     t.Run("returns NDJSON formatted logs", func(t *testing.T) {
	t.Setenv("AGENTFIELD_LOGS_ENABLED", "true")

	ring := newProcessLogRing(1024)
	ring.appendLine("stdout", "hello", false)

	req := httptest.NewRequest(http.MethodGet, "/agentfield/v1/logs", nil)
	req.Header.Set("Authorization", "Bearer")

	resp := httptest.NewRecorder()

	testAgentWithProcessLogRing(ring).handleAgentfieldLogs(resp, req)

	body := resp.Body.String()

	if !strings.Contains(body, "\n") {
		t.Fatalf("expected NDJSON format")
	}
   })
	  
}
